File: frontend/src/lib/crdts/AWMap.ts
import { DotContext } from "./DotContext";
import { DottedValue } from "./DottedValue";
import { AWSet } from "./AWSet";
import { ShoppingListItem } from "./ShoppingListItem";

/**
 * Add-Wins Map (AWMap) implementation using DotContext, AWSet, and DottedValue.
 * This structure is a conflict-free replicated data type (CRDT) that represents a map
 * where additions take precedence over removals in conflict resolution.
 */
export class AWMap<K, V extends { merge: (other: V) => void }> {
    private dotContext: DotContext;
    private localIdentifier: number;
    private keys: AWSet<K>;
    private values: Map<K, DottedValue<number, number, V>>;

    constructor(localIdentifier: number) {
        this.localIdentifier = localIdentifier;
        this.dotContext = new DotContext(localIdentifier);
        this.values = new Map<K, DottedValue<number, number, V>>();
        this.keys = new AWSet<K>(localIdentifier);
    }

    serializedMap(local: boolean = true) {
        let map = null;

        if (local) {
            map = new Map();
        } else {
            map = {};
        }

        for (const [key, value] of this.values) {
            if (local) {
                map.set(key, new DottedValue(value.identifier, value.event, value.value.toSerializable(local)));
            } else {
                map[key] = new DottedValue(value.identifier, value.event, value.value.toSerializable(local));
            }
        }

        return map;
    }

    getLocalIdentifier() {
        return this.localIdentifier;
    }

    setLocalIdentifier(localIdentifier: number) {
        this.localIdentifier = localIdentifier;
    }


    /**
     * Serializes the AWMap into a plain object for serialization or networking.
     */
    public toMessageProto(): {
        dotContext: { [key: number]: number };
        localIdentifier: { id: number };
        keys: { dotContext: { [key: number]: number }; localIdentifier: number; values: K[] };
        values: { [key: string]: { identifier: number; event: number; value: V } };
    } {
        const serializedValues: { [key: string]: { identifier: number; event: number; value: V } } = {};

        this.values.forEach((dottedValue, key) => {
            serializedValues[key.toString()] = dottedValue.toMessageDottedValue();
        });

        return {
            dotContext: this.dotContext.toMessageDotContext(),
            localIdentifier: this.localIdentifier.toMessageNodeIdentifier(),
            keys: this.keys.toMessageAWSet(),
            values: serializedValues,
        };
    }

    clone() {
        const cloned = new AWMap<K, V>(this.localIdentifier);
        cloned.setDotContext(this.dotContext.clone());
        cloned.setKeys(this.keys.clone());
        cloned.setValues(this.values);
        return cloned;
    }

    toSerializable(local: boolean = true) {
        return {
            "localIdentifier": this.localIdentifier,
            "values": this.serializedMap(local),
            "keys": this.keys.toSerializable(local),
            "dotContext": this.dotContext.toSerializable(local)
        }
    }

    /**
     * Retrieves the latest dot value for a given replica ID.
     */
    public latestDot(id: number): number | null {
        return this.dotContext.latestReplicaDot(id) || null;
    }

    public setDotContext(dotContext: DotContext): void {
        this.dotContext = dotContext;
    }

    public setKeys(keys: AWSet<K>): void {
        this.keys = keys;
    }

    public setValues(values: Map<K, DottedValue<number, number, V>>): void {
        this.values = values;
    }

    public getValues(): Map<K, DottedValue<number, number, V>> {
        return this.values;
    }

    public getDotContext(): DotContext {
        return this.dotContext;
    }

    public getValue(key: K): DottedValue<number, number, V> | undefined {
        return this.values.get(key);
    }

    public getKeys(): AWSet<K> {
        return this.keys;
    }

    /**
     * Adds a value to the map. If the key already exists, merges the values.
     */
    public add(id: K, value: V): void {
        const item = this.values.get(id);

        if (!item) {
            const dot = this.dotContext.nextOfReplica(this.localIdentifier);
            this.dotContext.getDots().set(this.localIdentifier, dot);
            this.values.set(id, {
                identifier: this.localIdentifier,
                event: dot,
                value: value
            });
            this.keys.add(id);
        } else {
            item.value.merge(value);
        }
    }

    /**
     * Removes a key-value pair from the map.
     */
    public remove(id: K): void {
        this.values.delete(id);
    }

    /**
     * Merges another AWMap into this one.
     */
    public merge(other: AWMap<K, V>): void {
        const localKeys = new Set(this.values.keys());
        const otherKeys = new Set(other.values.keys());

        // Find common keys and merge their values
        for (const key of localKeys) {
            if (otherKeys.has(key)) {
                const localValue = this.values.get(key);
                const otherValue = other.values.get(key);
                if (localValue && otherValue) {
                    localValue.value.merge(otherValue.value);
                }
            }
        }

        // Add keys unique to the other map
        for (const key of otherKeys) {
            if (!localKeys.has(key)) {
                const otherValue = other.values.get(key);
                if (otherValue) {
                    this.values.set(key, otherValue);
                }
            }
        }

        this.keys.merge(other.getKeys());

        // Merge DotContext
        this.dotContext.merge(other.getDotContext());
    }

    static fromDatabase(awMap, localId: number) {
        const cloned = new AWMap(awMap.localIdentifier);
        const map = new Map();

        if (awMap.values.entries) {
            for (const [key, value] of awMap.values.entries()) {
                map.set(key, new DottedValue(value.identifier, value.event, ShoppingListItem.fromDatabase(value.value, localId)));
            }
        } else {
            for (const [key, value] of Object.entries(awMap.values)) {
                map.set(key, new DottedValue(value.identifier, value.event, ShoppingListItem.fromDatabase(value.value, localId)));
            }
        }

        cloned.setValues(map);
        cloned.setDotContext(DotContext.fromDatabase(awMap.dotContext, localId));
        cloned.setKeys(AWSet.fromDatabase(awMap.keys, localId));

        return cloned;
    }
}

File: src/backend/src/main/java/feup/sdle/crdts/AWMap.java
package feup.sdle.crdts;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import feup.sdle.ShoppingList;
import feup.sdle.cluster.NodeIdentifier;
import feup.sdle.message.AWMapProto;
import feup.sdle.message.DottedValueProto;

import java.util.*;

public class AWMap<K, V extends CRDTSingleMergeable<V>> {
    @JsonProperty("dotContext")
    private DotContext dotContext;
    @JsonProperty("localIdentifier")
    private int localIdentifier;
    @JsonProperty("keys")
    private AWSet<K> keys;
    @JsonProperty("values")
    private HashMap<K, DottedValue<Integer, Integer, V>> values;

    @JsonCreator
    public AWMap(int localIdentifier, DotContext dotContext, HashMap<K, DottedValue<Integer, Integer, V>> values, AWSet<K> keys) {
        this.localIdentifier = localIdentifier;
        this.dotContext = dotContext;
        this.values = values;
        this.keys = keys;
    }

    public int getLocalIdentifier() {
        return localIdentifier;
    }

    public void setLocalIdentifier(int localIdentifier) {
        this.localIdentifier = localIdentifier;
    }

    public AWMap(int localIdentifier) {
        this.localIdentifier = localIdentifier;
        this.dotContext = new DotContext(this.localIdentifier);
        this.values = new HashMap<>();
        this.keys = new AWSet<>(this.localIdentifier);
    }

    public HashMap<K, DottedValue<Integer, Integer, V>> getValues() {
        return this.values;
    }

    public Optional<Optional<Integer>> latestDot(Integer id) {
        return Optional.of(this.dotContext.latestReplicaDot(id));
    }

    public void setNodeIdentifier(int nodeIdentifier) {
        this.localIdentifier = nodeIdentifier;
        this.keys.setLocalIdentifier(this.localIdentifier);
    }

    public void setDotContext(DotContext dotContext) {
        this.dotContext = dotContext;
    }

    public void setKeys(AWSet<K> keys) {
        this.keys = keys;
    }

    public void setValues(HashMap<K, DottedValue<Integer, Integer, V>> values) {
        this.values = values;
    }

    public DotContext getDotContext() {
        return this.dotContext;
    }

    public DottedValue<Integer, Integer, V> getValue(K key) {
        return this.values.get(key);
    }

    public AWSet<K> getKeys() {
        return this.keys;
    }

    public void add(K id, V value) {
        DottedValue<Integer, Integer, V> item = this.values.get(id);

        if(item == null) {
            Integer dot = this.dotContext.nextOfReplica(localIdentifier);
            this.dotContext.getDots().put(localIdentifier, dot);
            this.values.put(id, new DottedValue<>(this.localIdentifier, dot, value));
            this.keys.add(id);
        } else {
            item.value().merge(value);
        }
    }

    public void remove(K id) {
        DottedValue<Integer, Integer, V> item = this.values.get(id);
        if(item == null) return;

        this.values.remove(id);
        this.keys.remove(id);
    }

    // (𝑚, 𝑐) ⊔ (𝑚′, 𝑐′) = ({𝑘 ↦ → v(𝑘) | 𝑘 ∈ dom 𝑚 ∪ dom 𝑚′ ∧ v(𝑘) ≠ ⊥}, 𝑐 ∪ 𝑐′)
    // where v(𝑘) = fst ((𝑚[𝑘], 𝑐) ⊔ (𝑚′ [𝑘], 𝑐′))
    public void merge(AWMap<K, V> other) {
        Set<K> localKeys = this.values.keySet();
        Set<K> otherKeys = new HashSet<>(other.values.keySet());

        Set<K> commonKeys = new HashSet<>(localKeys);
        commonKeys.retainAll(otherKeys);

        for(K key : commonKeys) {
            this.values.get(key).value().merge(other.values.get(key).value());
        }

        otherKeys.removeAll(localKeys);
        for(K key: otherKeys) {
           this.values.put(key, other.values.get(key));
        }

        this.keys.merge(other.keys);
        this.dotContext.merge(other.dotContext);
    }

    public AWMapProto.AWMap toMessageProto() {
        var builder = AWMapProto.AWMap.newBuilder()
                .setDotContext(this.dotContext.toMessageDotContext())
                .setLocalIdentifier(this.localIdentifier)
                .setKeys(this.keys.toMessageAWSet());

        HashMap<String, DottedValueProto.DottedValue> protoEntries = new HashMap<>();

        for (Map.Entry<K, DottedValue<Integer, Integer, V>> entry : this.values.entrySet()) {
            protoEntries.put((String) entry.getKey(), entry.getValue().toMessageDottedValue());
        }

        return builder.putAllValues(protoEntries).build();
    }

    public static AWMap<String, ShoppingListItem> fromMessageAWMap(AWMapProto.AWMap msgAWMap) {
        AWMap<String, ShoppingListItem> awMap = new AWMap<>(msgAWMap.getLocalIdentifier());
        awMap.setDotContext(DotContext.fromMessageDotContext(msgAWMap.getDotContext()));

        HashMap<String, DottedValue<Integer, Integer, ShoppingListItem>> values = new HashMap<>();
        for (Map.Entry<String, DottedValueProto.DottedValue> entry : msgAWMap.getValuesMap().entrySet()) {
            values.put(entry.getKey(), DottedValue.fromMessageDottedValueSLI(entry.getValue()));
        }

        awMap.setValues(values);
        return awMap;
    }
}
